Add simple bash script to run tests as done by CI

strk opened the pull request:
See https://github.com/qgis/QGIS-Enhancement-Proposals/issues/244

Alessandro: does this work for you ? I'm sure it can be improved over time, I already left some TODO comments in the shell script

strk committed:
Add simple bash script to run tests as done by CI

See https://github.com/qgis/QGIS-Enhancement-Proposals/issues/244
Invoke with --help for usage syntax

Will build in build-ci/ to not mess with user build/ dir

elpaso commented on .ci/run_tests.sh:
```diff
@@ -0,0 +1,48 @@
+#echo "PWD is $PWD"
+#echo "QGIS_WORKSPACE is $QGIS_WORKSPACE"
+
+test -d build || {
```
why is this needed?

I think QGIS needs to be built inside the dependencies image, something like:

```
docker pull qgis/qgis3-build-deps:latest
docker tag qgis/qgis3-build-deps:latest qgis3-build-deps


docker run -t --name qgis_container \
                     -v $(pwd):/root/QGIS \
                     --env-file .docker/docker-variables.env \
                     --env PUSH_TO_CDASH=true \
                     --env WITH_QT5=true \
                     --env WITH_QT6=false \
                     --env WITH_QUICK=false \
                     --env WITH_3D=false \
                     --env PATCH_QT_3D=false \
                     qgis3-build-deps \
                     /root/QGIS/.docker/docker-qgis-build.sh
```

strk replied:
You are right, testing in docker requires also BUILDING in docker. Unfortunately both the builder and the test runner at the moment hard-code the `build/` directory which means it would clash with a pre-existing build/ directory. Maybe I'll make that parametrizable. Do you know why .docker/docker-qgis-build.sh uses ccache at all ? How are those caches supposed to be surviving multiple runs ?

strk replied:
I guess we can always improve this in a second stage, for now it'd be good enough to be able to run a script to run the same job run by the official CI in the same (dockerized) environment.

strk committed:
Parametrize both source and build dir via env for docker/ci tests

CTEST_SOURCE_DIR defaulting to /root/QGIS
CTEST_BUILD_DIR defaulting to /root/QGIS/build

strk commented:
Next problem: .ci/config.ctest hard-codes CTEST_BINARY_DIRECTORY to CTEST_BUILD_DIR/build

strk commented:
I'm confused by this .docker/docker-qgis-build.sh snippet:
```
#######
# Build
#######
echo "${bold}Building QGIS...${endbold}"
echo "::group::build"
ctest -VV -S /root/QGIS/.ci/config_build.ctest
echo "::endgroup::"
```

Is it normal that `ctest` is used to *build* ?

strk commented:
With the script in its current state I'm finally able to run the tests with the same environment as GitHub-Actions, running:
> .ci/run_tests.sh

The test are still not all passing, but most of them are:
```
99% tests passed, 3 tests failed out of 812

::set-output name=TESTS_FAILING::3
Total Test time (real) = 1100.00 sec

The following tests FAILED:
        314 - test_provider_eptprovider (Child aborted)
        316 - test_provider_pdalprovider (Failed)
        445 - PyQgsExternalStorageWebDAV (Failed)
```

Also, the run attempts to submit the results to the Dashboard, but it seems to fail in doing so and not realizing it is failing:
```
Submit files
   SubmitURL: http://
   Submit failed, waiting 30 seconds...
   Retry submission: Attempt 1 of 3
   Submit failed, waiting 30 seconds...
   Retry submission: Attempt 2 of 3
   Submit failed, waiting 30 seconds...
   Retry submission: Attempt 3 of 3
   Error when uploading file: /root/QGIS/build-ci/Testing/20220519-1149/Test.xml
   Error message was: 
   Problems when submitting via HTTP

::endgroup::
  Test results submitted to: https://tinyurl.com/yysvc7hu
```

Additionally, the output is full of github-action commands, like `::set-output name=CDASH_URL::https://tinyurl.com/yysvc7hu` which are most likely NOT needed when we run the testsuite from our console, so it may be a good idea to further simplify the run.

elpaso commented:
> Additionally, the output is full of github-action commands, like `::set-output name=CDASH_URL::https://tinyurl.com/yysvc7hu` which are most likely NOT needed when we run the testsuite from our console, so it may be a good idea to further simplify the run.

Can you look at --env PUSH_TO_CDASH if setting it to false prevents the submit?

strk commented:
On Fri, May 20, 2022 at 03:01:34AM -0700, Alessandro Pasotti wrote:
> > Additionally, the output is full of github-action commands, like `::set-output name=CDASH_URL::https://tinyurl.com/yysvc7hu` which are most likely NOT needed when we run the testsuite from our console, so it may be a good idea to further simplify the run.
> 
> Can you look at --env PUSH_TO_CDASH if setting it to false prevents the submit?

I've pushed that setting, but didn't seem to make a difference.
The .ci/config.ctest has the cdash submission hard-coded, looks to me.
